Add Gaudi Backend

baptistecolle opened the pull request:
## What does this PR do?
This PR integrates the Intel Gaudi backend into TGI's main codebase. Previously, we maintained with Intel a separate fork for Gaudi devices at https://github.com/huggingface/tgi-gaudi. By incorporating this as a backend within TGI, we significantly improve maintainability and reduce development overhead.

## Current Status and Implementation
The current tgi-gaudi fork has some API drift from standard TGI (divergence in the launcher, router, and client code)

This PR enables the Gaudi backend to work seamlessly with unmodified TGI components:
- Launcher
- Router
- Client
Now the Gaudi backend is only responsible for the Server aspect of TGI.

You can run the build the image for the Gaudi backend with `make -C backends/gaudi image`. For more information please refer to the Gaudi backend's README.md. 

To review the difference between this new server and the old one, the PR is structured with a first commit that just imported the server with zero modification from the tgi-gaudi fork. The next PRs then modify this server to make it work with the main TGI.

## Differences with the TGI-Gaudi fork server
There is one key behavioral difference compared to the tgi-gaudi fork:

The tgi-gaudi fork uses `--max-batch-total-tokens` for warmup, which isn't available in standard TGI. However, since `--max-batch-size` is available during warmup, we can compute the equivalent value:

```
max-batch-size = max-batch-total-tokens / max-total-tokens
```

### Migration Example
Instead of:
```bash
docker run ... tgi_gaudi ... --max-total-tokens 1024 --max-batch-total-tokens 8192
```

Use:
```bash
docker run ... tgi_gaudi ... --max-total-tokens 1024 --max-batch-size 8
```
Where `--max-batch-size` (8) = `--max-batch-total-tokens` (8192) / `--max-total-tokens` (1024)

## Validation and Performance
We've validated that both non-shared and shared meta-llama/Llama-3.1-8B-Instruct models function correctly. Performance benchmarks using https://github.com/huggingface/inference-benchmarker show comparable results:

| Metric | Gaudi Backend | TGI Upstream |
|--------|---------------|--------------|
| Token throughput (tokens/sec) | 125 | 126 |
| Time to first token (p50, ms) | 73,349  | 58,469  |

As expected, performance remains consistent since the underlying code is nearly identical across implementations.

## Next Steps
When those next steps are done, tgi-gaudi will be deprecated, and all future development for Intel Gaudi can happen in the Gaudi backend on the main TGI repo.
- Manually validate all models supported by the Gaudi backend (https://github.com/huggingface/tgi-gaudi). Unfortunately, there is no automated testing on tgi-gaudi for the moment.
- Add CI pipeline in TGI to publish the image under `ghcr.io/huggingface/text-generation-inference:3.1.0-gaudi`
- Create Gaudi-specific documentation in the TGI docs

## Future Improvements
- Refactor the Gaudi server code to remove unsupported models and not used code. The current implementation inherits unnecessary code from the upstream TGI NVIDIA backend. For example, the tgi-gaudi server currently has modeling code for idefics, but this is not supported on Gaudi hardware. A more targeted approach similar to the Neuron server in TGI or optimum-tpu would improve maintainability.
- Implement automated integration tests for the container, as right now, there is no automated test on tgi-gaudi. We could use AWS Gaudi1 instances (dl1.24xlarge) as they are easily available on AWS compared to Gaudi2.

regisss commented on backends/gaudi/server/text_generation_server/utils/version.py:
```diff
@@ -0,0 +1,12 @@
+from optimum.habana.utils import get_driver_version
+from packaging.version import Version
+
+MIN_TGI_GAUDI_SYNAPSE_VERSION = Version("1.16.0")
```
Hmm this may need to be updated to 1.19.0 :thinking:

regisss reviewed (commented):
I left several comments, mainly about removing files that we don't need in the Gaudi backend. There are probably more files that could be deleted (basically everything that is specific to cuda/rocm, e.g. many custom modelings for flash attention etc).

baptistecolle commented on backends/gaudi/Makefile:
```diff
@@ -0,0 +1,50 @@
+		docker run -it \
+		--runtime=habana \
+		-e HABANA_VISIBLE_DEVICES=all \
+		-e PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
```
Yes! I think we could simplify the docker command a lot. I was planning on doing it in a refactoring later but I can do some refactoring now already if preferred. 

Based on the 13 docker run commands in the TGI-Fork Readme. Here is the distribution

## Environment Variables
| Environment Variable | Count | Percentage |
|----------------------|-------|------------|
| `HABANA_VISIBLE_DEVICES=all` | 13/13 | 100% |
| `OMPI_MCA_btl_vader_single_copy_mechanism=none` | 13/13 | 100% |
| `ENABLE_HPU_GRAPH=true` | 13/13 | 100% |
| `LIMIT_HPU_GRAPH=true` | 13/13 | 100% |
| `USE_FLASH_ATTENTION=true` | 13/13 | 100% |
| `FLASH_ATTENTION_RECOMPUTE=true` | 13/13 | 100% |
| `TEXT_GENERATION_SERVER_IGNORE_EOS_TOKEN=true` | 10/13 | 76.9% |
| `PT_HPU_ENABLE_LAZY_COLLECTIVES=true` | 8/13 | 61.5% |

I adapted the code now to align with those. 
HABANA_VISIBLE_DEVICES is set to all (previously unset)
OMPI_MCA_btl_vader_single_copy_mechanism is set to none (previously unset)
ENABLE_HPU_GRAPH is unchanged (was previously set to true so keep it)
LIMIT_HPU_GRAPH is set True by default now (previously false)
USE_FLASH_ATTENTION is set True by default now (previously false)
FLASH_ATTENTION_RECOMPUTE is set True by default now (previously false)
PT_HPU_ENABLE_LAZY_COLLECTIVES is set to True if model is sharded. This is always the case when the model is sharded PT_HPU_ENABLE_LAZY_COLLECTIVES needs to be set to true.

I believe `TEXT_GENERATION_SERVER_IGNORE_EOS_TOKEN` might be problematic as an argument. Because it leads to the server crashing as the llm will continue decoding indefinitely. (I left more not on that in another comment tho)

regisss replied:
I was mainly commenting about `PT_HPU_ENABLE_LAZY_COLLECTIVES` as I'm not sure if it has any impact on single-device configurations. But if the latency is the same with and without it in a single-device deployment, then let's keep it.

baptistecolle commented on backends/gaudi/server/Makefile-vllm:
it is cuda specific, so we can remove it

baptistecolle commented on backends/gaudi/server/pyproject.toml:
```diff
@@ -0,0 +1,42 @@
+[tool.poetry]
+name = "text-generation-server"
+version = "2.0.4"
```
No, the tgi-gaudi server is based on the TGI server 2.0.4 for the moment. Right now, the TGI server is at [2.0.5-dev0](https://github.com/huggingface/text-generation-inference/blob/main/server/pyproject.toml). We currently have some differences that prevent us from being aligned with the new version

regisss replied:
My bad, I didn't realize the server had a different versioning

baptistecolle commented:
Thanks for the review @regisss and @yao-matrix

@regisss I wanted to do refactoring later on the server (cf. Future Improvements in PR description), but yeah, it's a great initiative to start it now! I also believe we can cut a lot of the stuff on the tgi-gaudi server. Should we continue more here, or would you leave it like that for another PR? 
@yao-matrix Maybe you also want to comment on that?

For example:
- do we support or use all the layers here? https://github.com/huggingface/tgi-gaudi/tree/habana-main/server/text_generation_server/layers
- Should we remove some models here as those are not in the supported model list? https://github.com/huggingface/tgi-gaudi/tree/habana-main/server/text_generation_server/models 
- Is there more stuff that can/should be removed to make the server leaner and thus easier to maintain?

baptistecolle commented on backends/gaudi/server/Makefile-awq:
We use autoawq however, right? 🤔

Because we install it in [TGI-gaudi](https://github.com/huggingface/tgi-gaudi/blob/17f0d57581f3ebd4cbcbfd624ef6cece472e3b0e/Dockerfile#L79) `BUILD_CUDA_EXT=0 pip install git+https://github.com/AutoGPTQ/AutoGPTQ.git@097dd04e --no-build-isolation && \` 

Because of this install, multiple other file are required such as Makefile-selective-scan and others 

Should I keep the installation of AutoGPTQ in the server or should it be removed?

regisss replied:
Yeah, as @yao-matrix said, we can keep autoawq as it is supported on Gaudi but has never been added to tgi-gaudi.

regisss commented:
@baptistecolle I think it's fine to leave it for another PR as it will be easier to review (this PR is pretty big already hehe).

Regarding your questions:
- I'm 99% sure we don't need any of the code in the `layers` folder
- I think we just need `causal_lm.py` and `vlm_causal_lm.py`. Basically, we need files that import from optimum-habana.
- Custom kernels we don't need. And probably some other things but we can do that later.

baptistecolle commented:
I implemented the new changes, namely removing redundant files from the server and making the habana and pytorch version parameters for easier maintenance. Still more can be done, but this should be done in a new PR. https://github.com/huggingface/text-generation-inference/pull/3055#issuecomment-2684566472

One small additional refactoring was done:
I simplified the docker run commands to make it more user-friendly based on that comment https://github.com/huggingface/text-generation-inference/pull/3055#discussion_r1970004795

For example, before we add a command like that 
```bash
docker run -p 8080:80 -v $volume:/data --runtime=habana -e PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
 -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none \
 -e  HF_TOKEN=$hf_token -e ENABLE_HPU_GRAPH=true -e LIMIT_HPU_GRAPH=true \
 -e USE_FLASH_ATTENTION=true -e FLASH_ATTENTION_RECOMPUTE=true --cap-add=sys_nice \
 --ipc=host ghcr.io/huggingface/tgi-gaudi:2.3.1 --model-id $model --sharded true \
 --num-shard 8 --max-input-tokens 1024 --max-total-tokens 2048
```

Now, it will be
```bash
docker run -p 8080:80 -v $volume:/data --runtime=habana 
 -e  HF_TOKEN=$hf_token --cap-add=sys_nice \
 --ipc=host ghcr.io/huggingface/tgi-gaudi:2.3.1 --model-id $model --sharded true \
 --num-shard 8 --max-input-tokens 1024 --max-total-tokens 2048
```

More or less, this specifies the arguments that are mainly the same for all the commands.

Again, I performed a sanity check and benchmarked the changes, and the performance is the same as that of the previous upstream implementation (token per second and time to first token)

Narsil reviewed (approved):
LGTM had some nits.
